core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1410,10 +1410,12 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     val scheme = new URI(schemeCorrectedPath).getScheme
     if (!Array("http", "https", "ftp").contains(scheme)) {
       val fs = hadoopPath.getFileSystem(hadoopConfiguration)
-      if (!fs.exists(hadoopPath)) {
-        throw new FileNotFoundException(s"Added file $hadoopPath does not exist.")
+      val isDir = try {
+        fs.getFileStatus(hadoopPath).isDirectory
+      } catch {
+        case f: FileNotFoundException =>
+          throw new FileNotFoundException(s"Added file $hadoopPath does not exist.").initCause(f)
```

Comment:
I think the theme of most of these changes is making one call to the HDFS NN rather than two. `getFileStatus` is the one call now. But it will signal that the file isn't found with an exception. It seems OK to me. Here the catch-throw is even pretty redundant.